src/sqlfluff/dialects/dialect_snowflake.py

```diff
@@ -3251,13 +3251,15 @@ class MergeNotMatchedClauseSegment(BaseSegment):
             "WHEN",
             "NOT",
             "MATCHED",
+            Sequence("AND", Ref("ExpressionSegment"), optional=True),
             "THEN",
```

Comment:
I would remove these two lines.

The point of a `match_grammar` being different than a `parse_grammar` is to allow that to match quickly without doing the full parse. I think the addition of `Expression` in this PR makes that a little more effort, for not much extra gain so think just matching on `WHEN NOT MATCHED` is sufficient.

```suggestion
```